src/common/btree.c

```diff
@@ -2585,74 +2651,97 @@ btr_root_del_rec(struct btr_context *tcx, struct btr_trace *trace, void *args)
 			/* have more than one record, simply remove the record
 			 * to be deleted.
 			 */
-			if (btr_has_tx(tcx))
-				btr_node_tx_add(tcx, trace->tr_node);
+			if (btr_has_tx(tcx)) {
+				rc = btr_node_tx_add(tcx, trace->tr_node);
+				if (rc != 0)
+					return rc;
+			}
 
-			btr_node_del_leaf_only(tcx, trace, true, args);
+			rc = btr_node_del_leaf_only(tcx, trace, true, args);
 		} else {
 
-			btr_node_destroy(tcx, trace->tr_node, args, NULL);
-			if (btr_has_tx(tcx))
-				btr_root_tx_add(tcx);
+			rc = btr_node_destroy(tcx, trace->tr_node, args, NULL);
+			if (rc != 0)
+				return rc;
+
+			if (btr_has_tx(tcx)) {
+				rc = btr_root_tx_add(tcx);
+				if (rc != 0)
+					return rc;
+			}
 
 			root->tr_depth	= 0;
 			root->tr_node	= BTR_NODE_NULL;
 
 			btr_context_set_depth(tcx, 0);
 			D_DEBUG(DB_TRACE, "Tree is empty now.\n");
 		}
-
 	} else {
 		/* non-leaf node */
 		D_DEBUG(DB_TRACE, "Delete child from the root, key_nr=%d.\n",
 			node->tn_keyn);
 
-		if (btr_has_tx(tcx))
-			btr_node_tx_add(tcx, trace->tr_node);
+		if (btr_has_tx(tcx)) {
+			rc = btr_node_tx_add(tcx, trace->tr_node);
+			if (rc != 0)
+				return rc;
+		}
+
+		rc = btr_node_del_child_only(tcx, trace, true);
+		if (rc != 0)
+			return rc;
 
-		btr_node_del_child_only(tcx, trace, true);
 		if (node->tn_keyn == 0) {
 			/* only has zero key and one child left, reduce
 			 * the tree depth by using the only child node
 			 * to replace the current node.
 			 */
-			if (btr_has_tx(tcx))
-				btr_root_tx_add(tcx);
+			if (btr_has_tx(tcx)) {
+				rc = btr_root_tx_add(tcx);
+				if (rc != 0)
+					return rc;
+			}
 
 			root->tr_depth--;
 			root->tr_node = node->tn_child;
 
 			btr_context_set_depth(tcx, root->tr_depth);
 			btr_node_set(tcx, node->tn_child, BTR_NODE_ROOT);
-			btr_node_free(tcx, trace->tr_node);
+			rc = btr_node_free(tcx, trace->tr_node);
 
-			D_DEBUG(DB_TRACE, "Shrink tree depth to %d\n",
-				tcx->tc_depth);
+			D_CDEBUG(rc != 0, DLOG_ERR, DB_TRACE,
+				 "Shrink tree depth to %d: rc = %d\n",
+				 tcx->tc_depth, rc);
 		}
 	}
+
+	return rc;
 }
 
 static int
 btr_delete(struct btr_context *tcx, void *args)
 {
 	struct btr_trace	*par_tr;
 	struct btr_trace	*cur_tr;
+	int			 rc = 0;
 
 	for (cur_tr = &tcx->tc_trace[tcx->tc_depth - 1];; cur_tr = par_tr) {
-		bool	bubble_up;
-
 		if (cur_tr == tcx->tc_trace) { /* root */
-			btr_root_del_rec(tcx, cur_tr, args);
+			rc = btr_root_del_rec(tcx, cur_tr, args);
 			break;
 		}
 
 		par_tr = cur_tr - 1;
-		bubble_up = btr_node_del_rec(tcx, par_tr, cur_tr, args);
-		if (!bubble_up)
+		rc = btr_node_del_rec(tcx, par_tr, cur_tr, args);
+		if (rc != 0) {
+			if (rc > 0)
```

Comment:
I'd rather store error code in btr_context, mixing error code and other return values is always hard to read for me.